File: C/heap_sort.c
#include <stdio.h>

void heap_sort(int [], int );
void heapify(int [], int , int );
void swap(int *, int *);
void print_array(int [], int );
int main() {
    int i, n, arr[100];
    printf("\nHow many elements? ");
    scanf("%d", &n);
    printf("\nEnter elements: ");
    for (i=0; i<n; i++)
        scanf("%d", &arr[i]);
    printf("\nOriginal Array: ");
    print_array(arr, n);
    heap_sort(arr, n);
    printf("\nSorted Array: ");
    print_array(arr, n);
}

void heap_sort(int arr[], int n) {
    int i;
    for (i = n/2; i >= 0; i--) {
        heapify(arr, n, i);
    }
    for (i = n-1; i>0 ; i--) {
        swap(&arr[0], &arr[i]);
        heapify(arr, i, 0);
    }
}

void heapify(int arr[], int n, int i) {
    // Find largest among root, left child and right child
    int largest = i;
    int left = 2 * i + 1;
    int right = 2 * i + 2;

    if (left < n && arr[left] > arr[largest])
        largest = left;

    if (right < n && arr[right] > arr[largest])
        largest = right;

    // Swap and continue heapifying if root is not largest
    if (largest != i) {
        swap(&arr[i], &arr[largest]);
        heapify(arr, n, largest);
    }
}

// swap elements a and b in the array
void swap(int *a, int *b) {
    int temp = *a;
    *a = *b;
    *b = temp;
}

// print the array
void print_array(int arr[], int n) {
    int i;
    for (i=0; i<n; i++)
        printf("%d ", arr[i]);  
}

/*
    Time Complexity : O(nlogn)
    Space Complexity : O(1)
    Input : 
        How many elements: 5
        Enter elements: 5 4 3 2 1
    Output : 
        Original Array: 5 4 3 2 1
        Sorted Array: 1 2 3 4 5
*/

File: Python/Heap Sort.py
def heapify(arr, n, i): 
	largest = i 
	l = 2 * i + 1	 # left = 2*i + 1 
	r = 2 * i + 2	 # right = 2*i + 2 

	'''
	See if left child of root exists and is 
	greater than root '''	
	if l < n and arr[i] < arr[l]: 
		largest = l 

	# See if right child of root exists and is 
	# greater than root 
	if r < n and arr[largest] < arr[r]: 
		largest = r 

	# Change root, if needed 
	if largest != i: 
		arr[i],arr[largest] = arr[largest],arr[i] # swap 

		# Heapify the root. 
		heapify(arr, n, largest) 

def heapSort(arr): 
	n = len(arr) 

	# Build a maxheap. 
	for i in range(n//2 - 1, -1, -1): 
		heapify(arr, n, i) 

	# One by one extract elements 
	for i in range(n-1, 0, -1): 
		arr[i], arr[0] = arr[0], arr[i] # swap 
		heapify(arr, i, 0) 

print('Enter a list of numbers')
arr = list(map(int,input().split()))
heapSort(arr) 
print('After performing heap sort')
for i in arr:
    print(i,end =' ')


'''
Time Complexity => O(nlogn)
Space complexity => O(1)
I/O =>

Enter a list of numbers
29 21 1 2 4 2 5 99 33 54 23
After performing heap sort
1 2 2 4 5 21 23 29 33 54 99

'''